ENH: make typing module available at runtime

person142 opened the pull request:
Closes https://github.com/numpy/numpy/issues/16550.

This makes `np.typing.ArrayLike` and `np.typing.DtypeLike` available
at runtime in addition to typing time. Some things to consider:

- `ArrayLike` uses protocols, which are only in the standard library
  in 3.8+, but are backported in `typing_extensions`. This
  conditionally imports `Protocol` and sets `_SupportsArray` to `Any`
  at runtime if the module is not available to prevent NumPy from
  having a hard dependency on `typing_extensions`. Since e.g. mypy
  already includes `typing_extensions` as a dependency, anybody
  actually doing type checking will have it set correctly.
- We are starting to hit the edges of "the fiction of the stubs". In
  particular, they could just cram everything into `__init__.pyi` and
  ignore the real structure of NumPy. But now that typing is available
  a runtime, we have to e.g. carefully import `ndarray` from `numpy`
  in the typing module and not from `..core.multiarray`, because
  otherwise mypy will think you are talking about a different
  ndarray. We will probably need to do some shuffling the stubs into
  more fitting locations to mitigate weirdness like this.

person142 committed:
MAINT: make typing module available at runtime

Closes https://github.com/numpy/numpy/issues/16550.

This makes `np.typing.ArrayLike` and `np.typing.DtypeLike` available
at runtime in addition to typing time. Some things to consider:

- `ArrayLike` uses protocols, which are only in the standard library
  in 3.8+, but are backported in `typing_extensions`. This
  conditionally imports `Protocol` and sets `_SupportsArray` to `Any`
  at runtime if the module is not available to prevent NumPy from
  having a hard dependency on `typing_extensions`. Since e.g. mypy
  already includes `typing_extensions` as a dependency, anybody
  actually doing type checking will have it set correctly.
- We are starting to hit the edges of "the fiction of the stubs". In
  particular, they could just cram everything into `__init__.pyi` and
  ignore the real structure of NumPy. But now that typing is available
  a runtime, we have to e.g. carefully import `ndarray` from `numpy`
  in the typing module and not from `..core.multiarray`, because
  otherwise mypy will think you are talking about a different
  ndarray. We will probably need to do some shuffling the stubs into
  more fitting locations to mitigate weirdness like this.

person142 commented on numpy/tests/typing/pass/array_like.py:
```diff
@@ -1,13 +1,7 @@
-    ArrayLike = Any
-    DtypeLike = Any
-    _SupportsArray = Any
+from numpy.typing import ArrayLike, DtypeLike, _SupportsArray
```
We also execute all the code in `pass`, so we're also testing here that you can really import these things at runtime.

person142 commented on numpy/typing/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from ._array_like import _SupportsArray, ArrayLike
```
There isn't much code in this package, but since typing is so verbose it would be a little painful to keep `Any`, `overload`, ... out of the public namespace if we crammed everything in a `typing.py` file. So instead make a package and use this init to make sure we're only exporting exactly what we mean to.

person142 commented on numpy/typing/_array_like.py:
```diff
@@ -0,0 +1,27 @@
+from numpy import ndarray
+from ._dtype_like import DtypeLike
+
+if sys.version_info >= (3, 8):
```
This is the "no hard dependency on `typing_extensions` dance mentioned in the PR description.

person142 replied:
I'd worry about that because it's pretty common for projects to run their tests suites with warnings turned into errors, and I could see something like e.g. a SciPy test run that doesn't have mypy installed and errors out because of the warning.

BvB93 replied:
Hmm, that might be a problem, yes.
Does NumPy have a logger where where such information could be displayed? Since I suspect that silently setting `_SupportsArray` to `Any` (and by extension `ArrayLike`) could result in some unexpected issues (at least from a end user perspective).

If not, then it should be mentioned in the documentation.

person142 replied:
No logger, when most of the action is happening in C code (potentially with the GIL released) logging doesn't work great. I added a big warning to to the top of the documentation in https://github.com/numpy/numpy/pull/16558/commits/4a120f0d073506329b564bfe9e69db2f6b612964.

BvB93 replied:
Ah, so be it.
At least it is documented now, which is the most important thing.

person142 commented on numpy/typing/_array_like.py:
```diff
@@ -0,0 +1,27 @@
+import sys
+from typing import Any, overload, Sequence, Tuple, Union
+
+from numpy import ndarray
```
This is an example of what I was trying to describe in the PR description where if we do `from numpy import ndarray`, then mypy goes and looks in `__init__.pyi`, finds `ndarray`, and types it correctly. But if we were to do `from ..core.multiarray import ndarray`, then it would find no stubs for that file and fall back on treating `ndarray` as `Any`, which would be bad.

person142 commented:
Now that I think of it, there is a much simpler solution to this problem-keep the current `typing.pyi` and define a `typing.py` file that just contains:

```python
from typing import _Any

DtypeLike = _Any
ArrayLike = _Any
```

That is, the types are correctly defined at typing time and unconditionally defined to be _whatever_ are runtime. It's short, avoids the "where do I import from" question, and the "is `typing_extensions` installed" question.

The drawback is that runtime introspection of the types is now impossible, but we mainly intended this as a syntactic convenience, i.e. we wanted people to be able to `from numpy.typing import ArrayLike` unconditionally versus having to do that in a `TYPE_CHECKING` block.

What do people think?

BvB93 commented:
> Now that I think of it, there is a much simpler solution to this problem-keep the current `typing.pyi` and define a `typing.py` file that just contains:

I like this solution, though I feel if we ever decide to expose a `Protocol` subclasses then those should also be available at run-time. 
Namely, this would allow one to use them for performing `isinstance()` and `issubclass()` checks.

Nevertheless, this is not an issue if we're just exposing `ArrayLike` and `DtypeLike`.

BvB93 commented:
> Now that I think of it, there is a much simpler solution to this problem-keep the current `typing.pyi` and define a `typing.py` file that just contains:

Though I do feel a value more descriptive than `Any` would be useful.

What about plain string?
``` python
DtypeLike = "numpy.typing.DtypeLike"
ArrayLike = "numpy.typing.ArrayLike"
```

person142 committed:
DOC: add documentation for the numpy.typing module

person142 commented:
Ok, I've added initial documentation for the `numpy.typing` module in https://github.com/numpy/numpy/pull/16558/commits/70130f848b7c526862fa6ff9667f078a628d86a1.

person142 committed:
DOC: add warning about typing-extensions module to numpy.typing docs

Typing `ArrayLike` correctly relies on `Protocol`, so warn users that
they should be on 3.8+ or install `typing-extensions` if they want
everything to work as expected.

person142 commented:
After all the discussion above, how do people feel about the approach taken here?

charris commented:
The release notes for typing work should probably go under `New features` rather than improvements and there should also be a release note for this PR. @seberg How does one label a release note fragment for two word section headers?

WarrenWeckesser commented:
> How does one label a release note fragment for two word section headers?

According to [README.rst](https://github.com/numpy/numpy/blob/master/doc/release/upcoming_changes/README.rst) in `doc/release/upcoming_changes`, its `#####.new_feature.rst`.

seberg commented:
The list of available sections is in the `pyproject.toml`.

person142 committed:
DOC: add release note about `np.typing` being available at runtime

person142 commented:
Release note added in https://github.com/numpy/numpy/pull/16558/commits/c88f5a232222b777a00e40648a6121428858df8d; opened https://github.com/numpy/numpy/pull/16603 to fix up the previous typing release note to be a new feature.

mattip commented:
A few questions, I am new to the world of typing. Feel free to point me to background reading if that is easier than answering the questions directly.

Does this slow down import time?

Could you give a higher-level picture of when availability of typing at runtime is desired, what use-case this answers? Is it typical for libraries (vs. user code) to provide such an ability?

person142 commented:
> Does this slow down import time?

It shouldn't in that it I opted not to add any `typing` import to the top level `__init__.py` under the assumption that (since Python typing is fairly verbose) people will generally be doing `from numpy.typing import ArrayLike` to keep the type annotations short. But, that's maybe a somewhat disingenuous answer, so here's the breakdown I'm getting from a

```
$ python -X importtime -c "import numpy.typing"
```

![numpy-typing-import-time](https://user-images.githubusercontent.com/15134881/84676730-b0c89a00-aee2-11ea-8e3d-29dd849b4ffc.png)

> what use-case this answers?

Right, so it's important to be clear that this doesn't enable anything new*; see e.g. https://github.com/numpy/numpy-stubs/pull/66#issuecomment-618784833 for a discussion of ways to use the things in `numpy.typing` without making them available at runtime. The argument from @shoyer in https://github.com/numpy/numpy-stubs/pull/66#issuecomment-618841857 was that it might be confusing for users that they can't do things like

```python
from numpy.typing import ArrayLike

x: ArrayLike = [1, 2, 3, 4]
```

*Though I will note that there are packages that do runtime introspection of types (pydantic); if anybody wanted to do something like that with `ArrayLike` then it would not be possible if it's only available at typing time.

> Is it typical for libraries (vs. user code) to provide such an ability?

I _think_ that it is atypical. It's hard to say why, some things that come to mind:

-  Not many packages have types yet, so there just aren't many examples to look at.
- For more class-based packages the classes already are the types, so they don't need to do anything extra to export those types. If NumPy had some array like ABC, then we would use that mechanism.
- NumPy is a lot more foundational than a lot of other packages.

If so inclined, @ethanhs might be able to offer better insights on this question.

shoyer commented on numpy/typing/_array_like.py:
```diff
@@ -0,0 +1,27 @@
+else:
+    _SupportsArray = Any
+
+ArrayLike = Union[bool, int, float, complex, _SupportsArray, Sequence]
```
It would be nice to support buffer protocols, but Python's typing doesn't support that yet. In the meantime I would suggest adding `memoryview` and a comment referencing https://github.com/python/typing/issues/593

person142 replied:
Since `memoryview`s are `Sequence`s we do allow them currently, e.g.

```
import numpy as np

x = b'foobar'
v = memoryview(x)
np.array(v)
```

passes mypy. I will add the comment though.

person142 replied:
We should add a test explicitly for memoryviews though; I'll open an issue.

ethanhs replied:
I'd highly recommend commenting on this B.P.O issue asking for support for a Buffer protocol in typing, as that is the better place than the typing issue: https://bugs.python.org/issue27501

shoyer commented on numpy/typing/__init__.py:
```diff
@@ -0,0 +1,65 @@
+The ``ArrayLike`` type tries to avoid creating object arrays. For
+example,
+
+.. code-block:: python
+
+  np.array(x**2 for x in range(10))
+
+is valid NumPy code which will create an object array. The types will
```
I think this example would be better motivated by showing the output, REPL style:
```
In [2]: np.array(x**2 for x in range(10))
   ...:
Out[2]: array(<generator object <genexpr> at 0x1118c5a20>, dtype=object)
```

Could also substitute "object array" -> "0-dimensional object array"

Most readers will probably not realize that this code works in an unexpected way otherwise!

shoyer commented on numpy/typing/__init__.py:
```diff
@@ -0,0 +1,65 @@
+
+  np.array(x**2 for x in range(10))
+
+is valid NumPy code which will create an object array. The types will
+complain about this usage however.
```
The language "the types will complain" sounds a little weird to me.

Types don't complain, they just are :).

Instead, I would say "Type checkers will complain"

shoyer commented on numpy/typing/__init__.py:
```diff
@@ -0,0 +1,65 @@
+
+is valid NumPy code which will create an object array. The types will
+complain about this usage however.
+
```
Can we add a brief note on the suggested work-around?

The obvious way would be to add a comment disabling typing:
```
  np.array(x**2 for x in range(10))  # type: ignore
```

Are there other recommended options?

I think we've also discussed making checks less strict if `dtype=object` is specified, e.g.,
```
  np.array(x**2 for x in range(10), dtype=object)
```
I don't know if that works yet. If it does, perhaps we should mention it, too.

person142 replied:
> Are there other recommended options?

The other way we test for:

https://github.com/numpy/numpy/blob/master/numpy/tests/typing/pass/array_like.py#L43)

is adding an explicit `Any` annotation. I've added examples of both methods to the docs.

> I don't know if that works yet. If it does, perhaps we should mention it, too.

Seems like @seberg would probably know the answer to that?

shoyer commented:
> > Does this slow down import time?
> 
> It shouldn't in that it I opted not to add any `typing` import to the top level `__init__.py` under the assumption that (since Python typing is fairly verbose) people will generally be doing `from numpy.typing import ArrayLike` to keep the type annotations short.

The overhead of importing the standard library's `typing` module is quite small, and it's increasingly likely that almost any non-trivial Python program will import `typing` at runtime -- soon likely including NumPy itself when we add type annotations.

So I think it should be fine to import `typing` at the top level `__init__.py`, so `np.typing.ArrayLike` works.

person142 committed:
DOC: clarify `ArrayLike` example in typing docs

person142 committed:
DOC: add note about supporting buffer protocols in `ArrayLike`

person142 commented:
Thanks for the review @shoyer, most points should be addressed in https://github.com/numpy/numpy/pull/16558/commits/c63f2333288772defcd84627986b035b6e7018ef and https://github.com/numpy/numpy/pull/16558/commits/347a368cd937b73c74ec8f684dcbaacb233ec84a. I didn't add the typing import to `__init__.py` yet though, since it seems likely other people will have opinions about that.

ethanhs commented:
Oh, weird, I wrote out a comment but I suppose I didn't hit the comment button.

Anyway, making types available at runtime should be the *default* in my opinion. There are a few reasons why:

1. 3.6 support. Python 3.6 not only will be supported until the very end of 2021 (https://devguide.python.org/#status-of-python-branches), it is also the most popular Python version for package downloads off PyPi (https://pypistats.org/packages/numpy, scroll down to "Daily Download Quantity of numpy Package"). You can somewhat get around this issue if you have people put quotes around their types or put everything in stubs, but that isn't very ergonomic.
2. Type aliases. `AType = Union[int, str, ndarray]`. These must be evaluated at runtime, or put in `if typing.TYPE_CHECKING` blocks, which is a bit ugly.
3. As previously mentioned, it is also less confusing to people newer to typing.

person142 committed:
DOC: add reference to Python issue about buffer protocols

person142 commented:
Thanks for that perspective @ethanhs.

Anybody else have strong opinions on importing `typing` in `__init__`? From the above graph it looks to be something like a 2.7% overhead (though as @shoyer noted, it's becoming increasingly likely that typing will have already been imported somewhere else, which would reduce that). I don't really have strong feelings on this point.

mattip commented:
Thanks @person142
